### ISSUE TITLE
Restructure PLR for speed

### ISSUE THREAD

**Issue by AlJacq** [AlJacq/LorenzRegression#17]
# Optimizing the Computation of the Derivative

## Problem Statement

We need to compute the derivative given by the following double summation:

$$
der(k) = \sum_{i=1}^n \sum_{j < i} \pi_i \pi_j (y_i - y_j) k(u_{ij}) \left( \frac{x_{ik} - x_{jk}}{h} \right)
$$

where:
- $u_{ij} = \frac{x_i^T \theta - x_j^T \theta}{h}$
- $\theta$ evolves throughout the algorithm
- $k(u_{ij})$ is a kernel function

Computing this derivative directly for all $(i, j)$ pairs can be extremely costly, especially for large $n$. Thus, we explore strategies to optimize this computation.

### Leveraging the Computation of $der_0(k)$
At the beginning of the algorithm, $\theta$ is initialized to a vector of zeros. This means that $u_{ij} = 0$ for all $i, j$ initially.

At iteration $m$, instead of recomputing $der_m(k)$ from scratch, we can express it as:

$$
der_m(k) = der_0(k) + \sum_{i=1}^n\sum_{j < i} \pi_i \pi_j (y_i - y_j) \left( \frac{x_{ik} - x_{jk}}{h} \right) \left( k(u_{ij}^m) - k(0) \right)
$$

###  Many Terms are Zero and Can Be Skipped

Not all terms in the double summation contribute to the final result. Specifically:
- **Ties in the response**. If $y_i = y_j$, then the term is zero, so it can be skipped. This might happen quite frequently with income data, (e.g. because of a substantial mass at zero-income).
- **Ties in the index**. If $x_i^T \theta = x_j^T \theta$, then $u_{ij}=0$ and the term can be skipped. This happens extremely frequently at the beginning of the algorithm.

Loop-skipping can be implemented via an early exit of the $(i,j)$ term. 

### Details on Skipping

Early exit is already costly in itself. There is a more refined way to proceed

For illustration consider that **observations are sorted in terms of the index**. We are at iteration $i$ and we want to avoid computing $j$ terms that are ties with $i$ in terms of the index. Instead of performing the loop with $j=0$ to $j=i-1$, we can adapt the end as follows

```
 int j_end = i - icum[i];
 if (j_end < 0) j_end = 0;
 for (j = 0; j <= j_end; j++)
```

The value `icum[i]` counts how many times the value `index[i]` has appeared up to that point. This induces a huge computation gain since we can disregard the useless $j$-terms without having to enter the $j$-loop. 

The same strategy could apply for ties in the response but it would necessitate to have **observations sorted in terms of the response**. Since we cannot have both the observations sorted in terms of response and index at the same time, we cannot use this strategy to disregard all useless terms upfront. 

The current version uses this strategy dynamically, either for the index or for the response, depending on which one would disregard more terms.

**Comment by AlJacq** [comment]
# Optimizing the Computation of the Loss

## Problem Statement

The problem is quite similar

$$
loss = \sum_{i=1}^n \sum_{j < i} \pi_i \pi_j (y_i - y_j) K(u_{ij})
$$

where $K(u_{ij})$ is the integrated kernel.

## Key Observations

###  Many Terms are Zero and Can Be Skipped
- If $y_i = y_j$, then the term is zero so it can be skipped.
- If $u_{ij} < -1$ then the term is zero so it can be skipped.

By implementing **loop-skipping**, we avoid unnecessary computations and significantly reduce processing time.

### Leveraging the Computation of $loss_0$
At the beginning of the algorithm, we have $K(u_{ij})=1/2$ for all $i,j$ since $u_{ij}=0$ everywhere. At iteration $m$, instead of recomputing $loss_m$ from scratch, we can express it as:

$$
loss_m = loss_0 + \sum_{i=1}^n\sum_{j < i} \pi_i \pi_j (y_i - y_j) \left( K(u_{ij}^m) - 1/2 \right)
$$

This means that whenever $u_{ij}^m = 0$, the term remains unchanged from the initial computation and can be skipped.

The rest is the same as for the derivative.

**Comment by AlJacq** [comment]
Other implementations improved drastically the speed of the computation of the derivative and loss 

- Switching from `arma::vec` to `std::vector` inside the double loop. 
- Defining as much as possible terms outside the loop.
